File: trackseriessaver/dataSource/trackseries.py
import requests
import json
import os
import time
import transaction
from persistent.dict import PersistentDict
import concurrent.futures

from ..entities import Serie, Season, Episode, SerieEncoder
from ..utils.network import get_url
from ..database import zodb
from ..utils.logger import logger
from ..utils.config import num_threads

BASE_URL = "https://api.trackseries.tv/v1"


def threaded_save_serie(access_token: str, id: int, image_path: str, name: str) -> Serie:
    """
    A threaded function to optimize the saving of the series

    Args:
        access_token (str): the access token of the user
        id (int): the id of the serie
        image_path (str): the path to the image folder

    Returns:
        Serie: the serie related to the id as given by the processSerie function
    """
    logger.info(f"Processing serie : {name}")
    return processSerie(get_serie(access_token, id), image_path)


def save_series(username: str, password: str, image_path: str):
    """
    A function that saves the series the user is following to a json file in the database, along with the images of the series
    """
    access_token: str = login(username, password)["access_token"]
    series = get_series(access_token)

    # prepare the folders
    if not os.path.exists(image_path):
        os.mkdir(image_path)
    # prepare the database
    if username not in zodb.dbroot["app_data"]:
        zodb.dbroot["app_data"][username] = PersistentDict()

    with concurrent.futures.ThreadPoolExecutor(max_workers=num_threads) as executor:
        start = time.time()
        futures = []
        for serie in series:
            id: int = serie["id"]
            future = executor.submit(threaded_save_serie, access_token, id, image_path, serie["name"])
            futures.append(future)

        for future, serie in zip(futures, series):
            processedSerie: Serie = future.result()
            id: int = serie["id"]
            zodb.dbroot["app_data"][username][id] = processedSerie
        transaction.commit()
        end = time.time()
        logger.info(f"Processing took {end - start} seconds")

    # print the first 5 series in the database into a json file
    for serie in list(zodb.dbroot["app_data"][username].values())[:5]:
        with open(f"tmp_data/{serie.id}.json", "w") as f:
            json.dump(serie, f, indent=4, cls=SerieEncoder)


def login(username: str, password: str) -> dict:
    """
    Login to the trackseries api. Returns a dict with the access_token and more info on the user

    Args:
        username (str): the username of the user
        password (str): the password of the user

    Returns:
        dict: a dict with the access_token and more info on the user. If the login failed, the dict will contain an error message under the key "message"
    """
    url: str = BASE_URL + "/Account/Login"
    payload = {"username": username, "password": password}
    headers = {"content-type": "application/json"}
    response = requests.post(url, data=json.dumps(payload), headers=headers)
    return response.json()


def get_series(access_token: str) -> list[dict]:
    """
    Get all the series the user is following

    Args:
        access_token (str): the access_token of the user

    Returns:
        list[dict]: a list of dicts with the info of the series the user is following
    """
    url: str = BASE_URL + "/Follow/Series"
    headers = {"Authorization": "Bearer " + access_token}
    response = get_url(url, headers=headers)
    return response.json()


def get_serie(access_token: str, serie_id: int) -> dict:
    """
    Get the info of a specific serie

    Args:
        access_token (str): the access_token of the user
        serie_id (int): the id of the serie

    Returns:
        dict: a list with the info of the serie, including the seasons and episodes with detailed info
    """
    url: str = f"{BASE_URL}/Follow/Series/{serie_id}/extended"
    headers = {"Authorization": "Bearer " + access_token}
    response = get_url(url, headers=headers)
    return response.json()


def save_image(url: str, path: str) -> None:
    """
    Save an image from a url to a file

    Args:
        url (str): the url of the image
        path (str): the path to save the image to
    """
    # if the image already exists, don't download it again
    if os.path.exists(path):
        return
    # get the image from the url
    img = get_url(url)
    # save it to a file
    with open(path, "wb") as f:
        f.write(img.content)


def processEpisodes(episodes: list[dict], imgPath: str) -> list[Episode]:
    result: list[Episode] = []
    for episode in episodes:
        serieId: int = episode["serieId"]
        id: int = episode["id"]
        directors: list[str] = episode["directors"]
        title: str = episode["title"]
        number: int = episode["number"]
        seasonNumber: int = episode["seasonNumber"]
        if episode["overview"] is None:
            overview: str = ""
        else:
            overview: str = episode["overview"].encode("utf-8").decode("utf-8")
        image: str = episode["image"]
        watched: bool = episode["watched"]
        imdbId: str = episode["ids"]["imdb"]
        tvdbId: int = episode["ids"]["tvdb"]
        airDate: str = episode["firstAired"]
        save_image(image, f"{imgPath}/{number}.jpg")
        image: str = f"{imgPath}/{number}.jpg"
        result.append(
            Episode(
                serieId=serieId,
                id=id,
                directors=directors,
                title=title,
                number=number,
                seasonNumber=seasonNumber,
                overview=overview,
                image=f"{imgPath}/{number}.jpg",
                watched=watched,
                imdbId=imdbId,
                tvdbId=tvdbId,
                airDate=airDate,
            )
        )
    return result


def processSeasons(seasons: list[dict], imagePath: str) -> list[Season]:
    result: list[Season] = []
    for season in seasons:
        number: int = season["seasonNumber"]
        # make the directory, if it doesn't exist
        seasonPath = imagePath + "/" + str(number)
        if not os.path.exists(seasonPath):
            os.mkdir(seasonPath)
        posterImage: str = season["poster"]
        save_image(posterImage, f"{seasonPath}/season{number}Poster.jpg")
        numEpisodes: int = season["progress"]["numEpisodes"]
        numEpisodesWatched: int = season["progress"]["numEpisodesWatched"]
        episodes: list[Episode] = processEpisodes(season["episodes"], seasonPath)
        result.append(
            Season(
                number=number,
                posterImage=f"{seasonPath}/season{number}Poster.jpg",
                numEpisodes=numEpisodes,
                numEpisodesWatched=numEpisodesWatched,
                episodes=episodes,
            )
        )
    return result


def processSerie(serie: dict, baseImagePath: str) -> Serie:
    id: int = serie["id"]
    name: str = serie["name"]
    numEpisodes: int = serie["progress"]["numEpisodes"]
    numEpisodesWatched: int = serie["progress"]["numEpisodesWatched"]
    numSeasons: int = len(serie["seasons"])
    overview: str = serie["overview"]
    imdbId: str = serie["imdbId"]
    tvdbId: int = serie["tvdbId"]
    posterImage: str = serie["images"]["poster"]
    bannerImage: str = serie["images"]["banner"]
    fanartImage: str = serie["images"]["fanart"]
    imagePath = baseImagePath + "/" + str(id)
    # make the directory, if it doesn't exist
    if not os.path.exists(imagePath):
        os.mkdir(imagePath)
    for img, fileName in [
        (posterImage, f"posterImage_{id}"),
        (bannerImage, f"bannerImage_{id}"),
        (fanartImage, f"fanartImage_{id}"),
    ]:
        save_image(img, f"{imagePath}/{fileName}.jpg")
    status: str = serie["status"]
    seasons: list[Season] = processSeasons(serie["seasons"], imagePath)
    return Serie(
        id=id,
        name=name,
        numEpisodes=numEpisodes,
        numEpisodesWatched=numEpisodesWatched,
        numSeasons=numSeasons,
        overview=overview,
        imdbId=imdbId,
        tvdbId=tvdbId,
        posterImage=f"{imagePath}/posterImage_{id}.jpg",
        bannerImage=f"{imagePath}/bannerImage_{id}.jpg",
        fanartImage=f"{imagePath}/fanartImage_{id}.jpg",
        status=status,
        seasons=seasons,
    )

File: backend/src/api/trackseries.ts
import { getURL } from "../utils/getUrl";
import { logger } from "../utils/logger";
import fs from "fs";
import {
  Serie,
  Season,
  Episode,
  IEpisode,
  ISeason,
  ISerie,
} from "../models/Serie";
import { User } from "../models/User";

const baseUrl: string = "https://api.trackseries.tv/v1";

async function threadedSaveSerie(
  accessToken: string,
  id: number,
  imagePath: string,
  name: string,
): Promise<ISerie> {
  /**
   * A threaded function to optimize the saving of the series
   *
   * @param access_token - the access token of the user
   * @param id - the id of the serie
   * @param image_path - the path to the image folder
   *
   * @return the serie related to the id as given by the processSerie function
   */

  logger.info(`Processing serie : ${name}`);
  return processSerie(await getSerie(accessToken, id), imagePath);
}

async function saveSeries(
  username: string,
  password: string,
  imagePath: string,
) {
  /**
   * A function that saves the series the user is following to a json file in the database, along with the images of the series
   *
   * @param username - the username of the user
   * @param password - the password of the user
   * @param image_path - the path to the image folder
   *
   * @return void
   *
   * @throws an error if the login failed
   *
   * @sideeffect saves the series to the database
   * @sideeffect saves the images of the series to the database
   * @sideeffect creates a folder for the images if it doesn't exist yet
   *
   **/
  const accessToken: string = login(username, password)["access_token"];
  const series = await getSeries(accessToken);

  // prepare the folders
  if (!fs.existsSync(imagePath)) {
    fs.mkdirSync(imagePath);
  }

  // start the processing
  const promises: Promise<ISerie>[] = [];
  for (const serie of series) {
    const id: number = serie.id;
    promises.push(threadedSaveSerie(accessToken, id, imagePath, serie.name));
  }
  const processedSeries: ISerie[] = await Promise.all(promises);

  // Saving series to the database
  const user = await User.findOne({ username }); // Find the user by username
  if (!user) {
    console.error("User not found.");
    return;
  }

  for (const processedSerie of processedSeries) {
    const serie = await Serie.findOne({ id: processedSerie.id }); // Find the serie by id
    // Add the saved serie to the user's series list
    user.series.push(serie._id); // Assuming that the Serie model returns an object with the _id field
  }
  await user.save(); // Save the user to the database

  logger.info("Finished saving series for user: " + username);
}

async function login(username: string, password: string): Promise<any> {
  /**
   * Login to the trackseries api. Returns a dict with the access_token and more info on the user
   *
   * @param username - the username of the user
   * @param password - the password of the user
   *
   * @return a dict with the access_token and more info on the user. If the login failed, the dict will contain an error message under the key "message"
   */

  const url: string = baseUrl + "/Account/Login";
  const payload = { username, password };
  const headers = { "content-type": "application/json" };
  fetch(url, {
    method: "POST",
    body: JSON.stringify(payload),
    headers,
  })
    .then((response) => {
      return response.json();
    })
    .then((data) => {
      return data;
    })
    .catch((error) => {
      logger.error(`Error in login: ${error}`);
    });
}

async function getSeries(accessToken: string): Promise<any> {
  /**
   * Get all the series the user is following
   *
   * @param access_token - the access_token of the user
   *
   * @return a list of dicts with the info of the series the user is following
   */

  const url: string = baseUrl + "/Follow/Series";
  const headers = { Authorization: "Bearer " + accessToken };
  const response = await getURL(url, headers);
  return response.data;
}

async function getSerie(accessToken: string, serieId: number): Promise<any> {
  /**
   * Get the info of a specific serie
   *
   * @param access_token - the access_token of the user
   * @param serie_id - the id of the serie
   *
   * @return a list with the info of the serie, including the seasons and episodes with detailed info
   */

  const url: string = baseUrl + "/Follow/Series/" + serieId + "/extended";
  const headers = { Authorization: "Bearer " + accessToken };
  const response = await getURL(url, headers);
  return response.data;
}

async function saveImage(url: string, filePath: string): Promise<void> {
  if (fs.existsSync(filePath)) {
    return;
  }
  const response = await getURL(url);
  fs.writeFileSync(filePath, response.data);
}

interface IEpisodeRaw {
  serieId: number;
  id: number;
  directors: string[];
  title: string;
  number: number;
  seasonNumber: number;
  overview: string;
  firstAired: string;
  image: string;
  watched: boolean;
  ids: {
    imdb: string;
    tvdb: number;
  };
}

async function processEpisodes(
  episodes: IEpisodeRaw[],
  imgPath: string,
): Promise<IEpisode[]> {
  const result: IEpisode[] = [];
  for (const episode of episodes) {
    const serieId: number = episode.serieId;
    const id: number = episode.id;
    const directors: string[] = episode.directors;
    const title: string = episode.title;
    const number: number = episode.number;
    const seasonNumber: number = episode.seasonNumber;
    const overview: string = episode.overview;
    const image: string = episode.image;
    const watched: boolean = episode.watched;
    const imdbId: string = episode.ids.imdb;
    const tvdbId: number = episode.ids.tvdb;
    const airDate: string = episode.firstAired;
    await saveImage(image, `${imgPath}/${number}.jpg`);
    const res = await Episode.create({
      serieId,
      id,
      directors,
      title,
      number,
      seasonNumber,
      overview,
      image: `${imgPath}/${number}.jpg`,
      watched,
      imdbId,
      tvdbId,
      airDate,
    });
    result.push(res);
  }
  return result;
}

interface ISeasonRaw {
  seasonNumber: number;
  poster: string;
  progress: {
    numEpisodes: number;
    numEpisodesWatched: number;
    numEpisodesAired: number;
    numEpisodesUnseen: number;
  };
  episodes: IEpisodeRaw[];
}

async function processSeasons(
  seasons: ISeasonRaw[],
  imagePath: string,
): Promise<ISeason[]> {
  const result: ISeason[] = [];
  for (const season of seasons) {
    const number: number = season.seasonNumber;
    // make the directory, if it doesn't exist
    const seasonPath = imagePath + "/" + String(number);
    if (!fs.existsSync(seasonPath)) {
      fs.mkdirSync(seasonPath);
    }
    const posterImage: string = season.poster;
    await saveImage(posterImage, `${seasonPath}/season${number}Poster.jpg`);
    const numEpisodes: number = season.progress.numEpisodes;
    const numEpisodesWatched: number = season.progress.numEpisodesWatched;
    const episodes: IEpisode[] = await processEpisodes(
      season.episodes,
      seasonPath,
    );
    const res = await Season.create({
      number,
      posterImage: `${seasonPath}/season${number}Poster.jpg`,
      numEpisodes,
      numEpisodesWatched,
      episodes,
    });
    result.push(res);
  }
  return result;
}

interface ISerieRaw {
  id: number;
  name: string;
  progress: {
    numEpisodes: number;
    numEpisodesWatched: number;
    numEpisodesAired: number;
    numEpisodesUnseen: number;
  };
  followers: number;
  firstAired: string;
  country: string;
  overview: string;
  runtime: number;
  status: string;
  network: string;
  airDay: string;
  airTime: string;
  contentRating: string;
  imdbId: string;
  tvdbId: number;
  language: string;
  images: {
    poster: string;
    fanart: string;
    banner: string;
  };
  seasons: ISeasonRaw[];
}

async function processSerie(
  serie: ISerieRaw,
  baseImagePath: string,
): Promise<ISerie> {
  const id: number = serie.id;
  const name: string = serie.name;
  const numEpisodes: number = serie.progress.numEpisodes;
  const numEpisodesWatched: number = serie.progress.numEpisodesWatched;
  const overview: string = serie.overview;
  const imdbId: string = serie.imdbId;
  const tvdbId: number = serie.tvdbId;
  const posterImage: string = serie.images.poster;
  const fanartImage: string = serie.images.fanart;
  const bannerImage: string = serie.images.banner;
  // make the directory, if it doesn't exist
  const imagePath = baseImagePath + "/" + String(id);
  if (!fs.existsSync(imagePath)) {
    fs.mkdirSync(imagePath);
  }
  await saveImage(posterImage, `${imagePath}/poster.jpg`);
  await saveImage(fanartImage, `${imagePath}/fanart.jpg`);
  await saveImage(bannerImage, `${imagePath}/banner.jpg`);
  const seasons: ISeason[] = await processSeasons(serie.seasons, imagePath);
  const result = await Serie.create({
    id,
    name,
    numEpisodes,
    numEpisodesWatched,
    numSeasons: seasons.length,
    overview,
    imdbId,
    tvdbId,
    posterImage: `${imagePath}/poster.jpg`,
    fanartImage: `${imagePath}/fanart.jpg`,
    bannerImage: `${imagePath}/banner.jpg`,
    seasons,
  });
  return result;
}

export { saveSeries };
